cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/worker/InMemoryWorkerRunner.java

```diff
@@ -0,0 +1,115 @@
+import co.cask.cdap.internal.app.runtime.SimpleProgramOptions;
+import co.cask.cdap.proto.ProgramType;
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.collect.HashBasedTable;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Table;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.inject.Inject;
+import org.apache.twill.api.RunId;
+import org.apache.twill.internal.RunIds;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+
+/**
+ * For running {@link Worker}. Only used in in-memory/standalone mode.
+ */
+public class InMemoryWorkerRunner extends AbstractInMemoryProgramRunner {
+  private static final Logger LOG = LoggerFactory.getLogger(InMemoryWorkerRunner.class);
+
+  @Inject
+  InMemoryWorkerRunner(ProgramRunnerFactory programRunnerFactory) {
+    super(programRunnerFactory);
+  }
+
+  @Override
+  protected ProgramOptions createComponentOptions(String name, int instanceId, int instances, RunId runId,
+                                                  Arguments userArguments) {
+    Map<String, String> options = ImmutableMap.of(ProgramOptionConstants.INSTANCE_ID, Integer.toString(instanceId),
+                                                  ProgramOptionConstants.INSTANCES, Integer.toString(instances),
+                                                  ProgramOptionConstants.RUN_ID, runId.getId());
+    return new SimpleProgramOptions(name, new BasicArguments(options), userArguments);
+  }
+
+  @Override
+  public ProgramController run(Program program, ProgramOptions options) {
+    // Extract and verify parameters
+    ApplicationSpecification appSpec = program.getApplicationSpecification();
+    Preconditions.checkNotNull(appSpec, "Missing application specification.");
+
+    ProgramType type = program.getType();
+    Preconditions.checkNotNull(type, "Missing processor type.");
+    Preconditions.checkArgument(type == ProgramType.WORKER, "Only WORKER process type is supported.");
+
+    WorkerSpecification workerSpec = appSpec.getWorkers().get(program.getName());
+    Preconditions.checkNotNull(workerSpec, "Missing WorkerSpecification for %s", program.getName());
+
+    //RunId for worker
+    RunId runId = RunIds.generate();
+    Table<String, Integer, ProgramController> components = startWorkers(program, runId, options.getUserArguments(),
+                                                                        workerSpec);
+    return new InMemoryProgramController(components, runId, program, workerSpec, options.getUserArguments(),
+                                         ProgramRunnerFactory.Type.WORKER_COMPONENT);
+  }
+
+  private Table<String, Integer, ProgramController> startWorkers(Program program, RunId runId, Arguments arguments,
```

Comment:
It will become easier to refactor that once we remove ServiceWorkers. I will add a TODO regarding the same.